subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/AbstractGroovyCompileAvoidanceIntegrationSpec.groovy

```diff
@@ -215,4 +215,174 @@ public @interface MyAnnotation { }
         executedAndNotSkipped ":b:compileGroovy"
         failure.assertHasCause('Bad AST transformation!')
     }
+
+    def "doesn't recompile when private element of implementation class changes"() {
+        given:
+        buildFile << """
+            project(':b') {
+                dependencies {
+                    implementation project(':a')
+                }
+            }
+        """
+        def sourceFile = file("a/src/main/${language.name}/ToolImpl.${language.name}")
+        sourceFile << """
+            public class ToolImpl { 
+                private String thing() { return null; }
+                private ToolImpl t = this;
+            }
+        """
+        file("b/src/main/${language.name}/Main.${language.name}") << """
+            public class Main { ToolImpl t = new ToolImpl(); }
+        """
+
+        when:
+        succeeds ":b:${language.compileTaskName}"
+
+        then:
+        executedAndNotSkipped ":a:${language.compileTaskName}"
+        executedAndNotSkipped ":b:${language.compileTaskName}"
+
+        when:
+        // change signatures
+        sourceFile.text = """
+            public class ToolImpl { 
+                private Number thing() { return null; }
+                private Object t = this;
+            }
+"""
+
+        then:
+        succeeds ":b:${language.compileTaskName}"
+        executedAndNotSkipped ":a:${language.compileTaskName}"
+        skipped ":b:${language.compileTaskName}"
+
+        when:
+        // add private elements
+        sourceFile.text = """
+            public class ToolImpl { 
+                private Number thing() { return null; }
+                private Object t = this;
+                private static void someMethod() { }
+                private String s;
+            }
+"""
+
+        then:
+        succeeds ":b:${language.compileTaskName}"
+        executedAndNotSkipped ":a:${language.compileTaskName}"
+        skipped ":b:${language.compileTaskName}"
+
+        when:
+        // remove private elements
+        sourceFile.text = """
+            public class ToolImpl { 
+            }
+"""
+
+        then:
+        succeeds ":b:${language.compileTaskName}"
+        executedAndNotSkipped ":a:${language.compileTaskName}"
+        skipped ":b:${language.compileTaskName}"
+
+        when:
+        // add public method, should change
+        sourceFile.text = """
+            public class ToolImpl { 
+                public void execute() { String s = toString(); } 
+            }
+"""
```

Comment:
Nit: Can you let the closing `"""` have the same indentation as the opening ones?

Reply:
Done